docs/docsite/rst/dev_guide/testing/sanity/import.rst

```diff
@@ -3,57 +3,102 @@ import
 
-       try:
-           import another_library
-       except ImportError:
-           HAS_ANOTHER_LIBRARY = False
-           ANOTHER_LIBRARY_IMPORT_ERROR = traceback.format_exc()
-       else:
-           HAS_ANOTHER_LIBRARY = True
+   import traceback
 
+   from ansible.module_utils.basic import missing_required_lib
 
-       # Later in module code:
+   try:
+       import another_library
+   except ImportError:
+       HAS_ANOTHER_LIBRARY = False
+       ANOTHER_LIBRARY_IMPORT_ERROR = traceback.format_exc()
+   else:
+       HAS_ANOTHER_LIBRARY = True
 
-       module = AnsibleModule(...)
+.. note::
 
-       if not HAS_ANOTHER_LIBRARY:
-           # Needs: from ansible.module_utils.basic import missing_required_lib
-           module.fail_json(
-               msg=missing_required_lib('another_library'),
-               exception=ANOTHER_LIBRARY_IMPORT_ERROR)
+   The ``missing_required_lib`` import above will be used below.
 
-2. In plugins:
+Then in the module code:
 
-   .. code-block:: python
+.. code-block:: python
 
-       # Instead of 'import another_library', do:
+   module = AnsibleModule(...)
 
-       from ansible.module_utils.six import raise_from
+   if not HAS_ANOTHER_LIBRARY:
+       module.fail_json(
+           msg=missing_required_lib('another_library'),
+           exception=ANOTHER_LIBRARY_IMPORT_ERROR)
 
-       try:
-           import another_library
-       except ImportError as imp_exc:
-           ANOTHER_LIBRARY_IMPORT_ERROR = imp_exc
-       else:
-           ANOTHER_LIBRARY_IMPORT_ERROR = None
+In plugins
+^^^^^^^^^^
 
+Instead of using ``import another_library``:
 
-       # Later in plugin code, for example in __init__ of the plugin:
+.. code-block:: python
 
-       if ANOTHER_LIBRARY_IMPORT_ERROR:
-           raise_from(
-               AnsibleError('another_library must be installed to use this plugin'),
-               ANOTHER_LIBRARY_IMPORT_ERROR)
-           # If you target only newer Python 3 versions, you can also use the
-           # 'raise ... from ...' syntax.
+   try:
+       import another_library
+   except ImportError as imp_exc:
+       ANOTHER_LIBRARY_IMPORT_ERROR = imp_exc
+   else:
+       ANOTHER_LIBRARY_IMPORT_ERROR = None
+
+Then in the plugin code, for example in ``__init__`` of the plugin:
+
+.. code-block:: python
+
+   if ANOTHER_LIBRARY_IMPORT_ERROR:
+       raise AnsibleError('another_library must be installed to use this plugin') from ANOTHER_LIBRARY_IMPORT_ERROR
+
+When used as base classes
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. important::
+
+   This solution builds on the previous two examples.
+   Make sure to pick the appropriate one before continuing with this solution.
+
+Sometimes an import is used in a base class, for example:
+
+.. code-block:: python
+
+   from another_library import UsefulThing
+
+   class CustomThing(UsefulThing):
+       pass
+
+One option is make the entire class definition conditional:
+
+.. code-block:: python
+
+   if not ANOTHER_LIBRARY_IMPORT_ERROR:
```

Comment:
isn't the point of the whole import shenanigans to provide meaningful error for the end user? If the class is partially or not defined at all it would create much more obscure error messages to the user hence more workload for support.

I don't know enough to propose any alternatives though, so this just my 2c, thinking aloud.

Reply:
You should never get to a point where you use this class.  The goal would be to error with a friendly message earlier than using the class.  As such, you have to avoid anything that would attempt to use the class in any meaningful way from the global scope of the module.

Reply:
Thanks for explaining ,makes sense!

If i get it correctly, then sanity tests would catch some module using this `CustomThing` without a proper checks anyway, bc it sanity tests run without the dependencies as well.

/resolved

Reply:
I'm assuming second example, where base class is mocked, unit tests would catch that?

Reply:
Unit tests should be running with the dependencies installed, so should never see the mocked class. The mocked class should only be present for sanity tests and end-users running with missing dependencies.

Reply:
There's no "gap" as long as the recommended error handling is included.